# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/mbc/ai/m4/datasets/spam-data.csv](https://static.bc-edx.com/mbc/ai/m4/datasets/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive-beta.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/mbc/ai/m4/datasets/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

I believe the Random Forests Classifier will perform best because:
* There are 57 featuers in this data set. This will be prone to overfitting
* Random Forests Classifier will make use of all the data, creating an ensemble of decision trees for the weak learners
* Random Forests is rubusts to outliers and runs efficiently on large data sets

## Split the Data into Training and Testing Sets

In [3]:
# Create the labels set `y` and features DataFrame `X`

# Features data
X = data.copy()
X.drop(
  "spam",
  axis=1,
  inplace=True
)

# display
# X.head()

# create the y data set
y = data["spam"]

#display
# y[:10]


In [4]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

0    2788
1    1813
Name: spam, dtype: int64

In [5]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [6]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()

In [7]:
# Fit the Standard Scaler with the training data
scaler.fit(X_train)


StandardScaler()

In [8]:
# Scale the training data
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

#display
# X_train_scaled[:10]

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [9]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

# instantiate the model
logistic_regression_model = LogisticRegression()


# train the model
logistic_regression_model.fit(X_train_scaled, y_train)

# score the model
print(f"Training Data Score: {logistic_regression_model.score(X_train_scaled, y_train)}")



Training Data Score: 0.9257971014492754


In [10]:
# Make and save testing predictions with the saved logistic regression model using the test data

# generate predictions from the model we just fit
testing_predictions = logistic_regression_model.predict(X_test_scaled)

# Review the predictions
# convert those predictions (and actual values) to a DF
results_df = pd.DataFrame({"Prediction": testing_predictions, "Actual": y_test})

#display
results_df

Prediction  Actual
3683           0       0
4412           0       0
2584           0       0
69             1       1
1844           0       0
...          ...     ...
1912           0       0
1260           1       1
801            1       1
2698           0       0
2095           0       0

[1151 rows x 2 columns]

In [11]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
logistics_accuracy_score = accuracy_score(y_test, testing_predictions)

print(f"Logistics Regressions Testing Data Score: {logistics_accuracy_score:.3%}")



Logistics Regressions Testing Data Score: 92.268%


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [12]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier

# unscaled data sets are: X_train, X_test, y_train and y_test

# create a random forest model instance
rf_model = RandomForestClassifier(n_estimators=128, random_state=78)

# fit the model
rf_model = rf_model.fit(X_train, y_train)

# make predicitons using the testing data
rf_predictions = rf_model.predict(X_test)

# calculate the accuracy score
rf_acc_score = accuracy_score(y_test, rf_predictions)

# display
print(f"Random Forests Accuracy Score; {rf_acc_score:.3%}")


Random Forests Accuracy Score; 95.830%


In [13]:
# Make and save testing predictions with the saved logistic regression model using the test data
logistics_accuracy_score = accuracy_score(y_test, testing_predictions)


# Review the predictions
print(f"Logistics Regressions Testing Data Score: {logistics_accuracy_score:.3%}")


Logistics Regressions Testing Data Score: 92.268%


In [14]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
rf_acc_score = accuracy_score(y_test, rf_predictions)
print(f"Random Forests Accuracy Score: {rf_acc_score:.3%}")


Random Forests Accuracy Score: 95.830%


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

As expected, the Random Forests model outperformed the Logistics Regression model
* Logistics Regressions Testing Data Score: 92.268%
* Random Forests Accuracy Score: 95.830% 

The results are as I originally expected. The data set that we used has 57 features. It would be worth exploring whether pre-processing the data with a tool like PCA to reduce the dimensionality would imrove the performance of the Logistics Regression model. While this *may* improve model performance, a drawback of this approach would be the loss of information of which features are imortant.